Search Marquee LCP Performance

JingleH opened the pull request:
Move parts of the block's decoration out of the eager phase and dedup a few network fetches to speed up LCP. This might seem trivial as I can only consistently measure a 0.1-0.2 second improvement but should still be good to have. We will resolve Template pages' TBT issues with the template-x revamp initiative.

**There should be no Visual/Functional changes to any pages.**

Resolves: https://jira.corp.adobe.com/browse/MWPW-154148

Before vs after (exact numbers can be fluctuating and I am not throttling on my local machine, but the left-shift of things especially network should be visible, thanks to deduping requests to save bandwidth, especially placeholders.json):
![SC 2024-07-10 at 2 57 37 PM](https://github.com/adobecom/express/assets/32369333/dd132a44-bb75-4f86-b0ca-fcb6d181be38)


Test URLs:
- Before: https://stage--express--adobecom.hlx.live/express/templates/flyer?martech=off
- After: https://search-marquee-lcp--express--adobecom.hlx.live/express/templates/flyer?martech=off

fullcolorcoder reviewed (changes requested):
There is zero test coverage for this file. It would be good if you could cover at least the lines you added.

JingleH commented on express/blocks/search-marquee/search-marquee.js:
```diff
@@ -340,23 +335,20 @@ function decorateLinkList(block) {
-
+async function lazyWork(block) {
+  const searchBarWrapper = await decorateSearchFunctions(block);
+  await buildSearchDropdown(block, searchBarWrapper);
```
sad things is that they need to happen in sequence... for example decorateLinkList() needs to happen at the end. We can refactor them a bit later so more work can be done in sequence. I am envisioning a major carousel refactor down the line too...

JingleH replied:
in fact, i noticed that separating parts of the decoration to lazily done, makes unit testing that part of decoration harder, so lazy me decided to just do them eagerly and synchronously now. This shouldn't compromise much of the gains we're expecting. Check if it makes sense @vhargrave! I'm still keeping the decorateLlinkList()/CKG part lazy tho

vhargrave replied:
I can take a proper look next week, but on the fly I would say, that I would rather have no unit test than compromising on any gains that we have tbh . My opinion is to keep is as asyncronous as possible mate 🙏

JingleH commented on express/scripts/utils.js:
```diff
@@ -1389,24 +1389,31 @@ export async function loadTemplateScript() {
  */
 
 export async function fetchPlaceholders() {
+  if (window.placeholders) return window.placeholders;
```
yeah @vhargrave, that would be my preferred approach too. Just before I go ahead, there are just 2 things for you and @fullcolorcoder:
1. Our placeholders are done differently than in Milo, and putting it in window does allow us to mock it more easily in unit tests. 
2. I believe almost every single page and LCP element depends on placeholders. How about we just always do a fetch in scripts.js and store it somewhere, and let other blocks can just grab it? This can turn a lot of async functions back to sync, since it only takes one `await fetchPlaceholders()` to flip a function's color. If we worry about evil blocks modifying placeholders object we can Object.freeze it too.

vhargrave replied:
hmm.. I don't mind the idea, but unfortunately this will only be a short term gain anyway. Fetching placeholders in milo is always asyncronous, so if we make it syncronous now we'll have to change it while migrating unfortuantely anyway 😢 
And from a performance perspective there would be arguments that could be made that placeholders should not be fetched right away, because LCP might not need, so the question is when is it appropriate and that's a whole rabbit hole to go down.

JingleH replied:
Yeah let's keep the idea unchanged. This way we can avoid a lot of changes in the repo too. It also looks like a lot of our unit tests already depend on that window.placeholders for mocking, so let's keep the task for turning it into a module-level for later, as that change will be less related to this PR

JingleH commented:
> There is zero test coverage for search marquee. It would be good if you could cover at least the lines you add.

@fullcolorcoder  you caught me good, sir! I added validation for the basic decorations and it should cover all the changes I'm making in this PR. To get to higher coverage for this block, we would need to rewrite/refactor/expose/export the original code a bit to improve its testability so that we can more reliably test those delayed work and interactions. It would be less trivial so would you be still approving now if I log a ticket and do that later? This whole block along with template-x (mega variant) are planned to be totally revamped and rewritten in either preact or web components later, so part of me just want to move on as early as possible...
